docs/connection.rst

```diff
@@ -56,11 +64,21 @@ The `!Connection` class
     .. autoattribute:: closed
         :annotation: bool
 
-    .. automethod:: cursor
 
-        Calling the method without a *name* creates a client-side cursor,
-        specifying a *name* crates a server-side cursor. See
-        :ref:`cursor-types` for the details.
+    .. method:: cursor(*, binary: bool = False, row_factory: Optional[RowFactory] = None) -> Cursor
+    .. method:: cursor(name: str, *, binary: bool = False, row_factory: Optional[RowFactory] = None) -> ServerCursor
```

Comment:
I'm not sure about the intent of this change. As is, it seems to me that `conn.cursor(row_factory=None)` does not work anymore as a way to reset the row factory that's been defined in `connect()` or `conn.row_factory`.

For instance, this crashes:
```diff
diff --git a/tests/test_connection.py b/tests/test_connection.py
index 8ab62ee..8290c8b 100644
--- a/tests/test_connection.py
+++ b/tests/test_connection.py
@@ -498,7 +498,7 @@ def test_row_factory(dsn):
         cur.execute("select 1, 1, 2")
         assert cur.fetchall() == [{1, 2}]
 
-    with conn.cursor(row_factory=tuple_row) as cur:
+    with conn.cursor(row_factory=None) as cur:
         cur.execute("select 1, 1, 2")
         assert cur.fetchall() == [(1, 1, 2)]
```

Reply:
Yes, it crashes but just because it uses `my_row_factory`, which assumes values to be string. It fails exactly the same way as:
```diff
@@ -498,7 +498,7 @@ def test_row_factory(dsn):
         cur.execute("select 1, 1, 2")
         assert cur.fetchall() == [{1, 2}]
 
-    with conn.cursor(row_factory=tuple_row) as cur:
+    with conn.cursor() as cur:
         cur.execute("select 1, 1, 2")
         assert cur.fetchall() == [(1, 1, 2)]
```

The sense of that change is to give `None` the meaning of default: in `connect()` it means `tuple_row`, in `conn.cursor()` it means `conn.row_factory`. The value `None` itself is no more valid as `row_factory` attribute, which is no more `Optional`.

So now there is an explicit `tuple_row` which represents the default behaviour. It behaves _like_ `lambda cur: lambda t: t`, but it is actually defined as `lambda cur: None`, because `None` is valid as `Transformer.make_row`. This avoids the need of having a sentinel object and having the `conn.row_factory` non optional makes the paths a bit simpler - no if at every use. As performance is probably influential, calling a function returning a constant at every query is not a heavy thing as it would calling a function on any row (as it would be the case if the default row factory was `lambda cur: lambda t: t`. 

I had to relax the `RowFactory` signature to allow it to return `None`. But now that I am looking at the code we could also make `Transform.make_row` non-nullable and just define:

```python
# row.py

def tuple_factory(cur: Cursor) -> Callable[[Sequence[Any]], Tuple[Any, ...]]:
    return tuple
```

So they python transformer can just use it
```diff
@@ -38,7 +38,7 @@ class Transformer(AdaptContext):
     __module__ = "psycopg3.adapt"
     _adapters: "AdaptersMap"
     _pgresult: Optional["PGresult"] = None
-    make_row: Optional[RowMaker] = None
+    make_row: RowMaker = tuple
 
     def __init__(self, context: Optional[AdaptContext] = None):
 
@@ -174,17 +174,13 @@ class Transformer(AdaptContext):
 
         records: List[Row]
         records = [None] * (row1 - row0)  # type: ignore[list-item]
-        if self.make_row:
-            mkrow = self.make_row
-        else:
-            mkrow = tuple
         for row in range(row0, row1):
             record: List[Any] = [None] * self._nfields
             for col in range(self._nfields):
                 val = res.get_value(row, col)
                 if val is not None:
                     record[col] = self._row_loaders[col](val)
-            records[row - row0] = mkrow(record)
+            records[row - row0] = self.make_row(record)
 
         return records
``` 

and in the Pyrex code, to skip the no-op call, it can be:
```diff
index ce88491..261bde9 100644
--- a/psycopg3_c/psycopg3_c/_psycopg3/transform.pyx
+++ b/psycopg3_c/psycopg3_c/_psycopg3/transform.pyx
@@ -385,7 +385,7 @@ cdef class Transformer:
             Py_INCREF(pyval)
             PyTuple_SET_ITEM(record, col, pyval)
 
-        if self.make_row:
+        if self.make_row is not tuple:
             return self.make_row(record)
         return record
```

What do you think?

Reply:
> The sense of that change is to give None the meaning of default [...]

I see now, thanks for explaining. It makes sense.

> But now that I am looking at the code we could also make Transform.make_row non-nullable and just define:
> ```python
> # row.py
> def tuple_factory(cur: Cursor) -> Callable[[Sequence[Any]], Tuple[Any, ...]]:
>     return tuple
> ```

This looks better, indeed.

Reply:
Done, and also made the attribute writable on the cursor